### DISCUSSION TITLE
VI_ERROR_TMO (-1073807339) when printing buffer data

### DISCUSSION THREAD

**Discussion by vuhoang8** [tektronix/tm_devices#274] [Q&A] [answered]
I'm currently writing a script for a pulse train on the Keithley 2612B and have run into a timeout error when attempting to print the readings from the two designated buffers.

I created the buffers using the code below:
```
smua = inst.commands.smu["a"]
    
ibuffer = inst.commands.buffer_var["smua.nvbuffer1"]
vbuffer = inst.commands.buffer_var["smua.nvbuffer2"]
    
inst.commands.reset()
    
ibuffer.clear()
ibuffer.appendmode = 1
ibuffer.collecttimestamps = 1
    
vbuffer.clear()
vbuffer.appendmode = 1
vbuffer.collecttimestamps = 1
```

Afterwards, I created the pulse train using an example adapted from the 2600B Reference Manual, page 3-49. However, when I run the following code:

`inst.print_buffers("smua.nvbuffer2.readings", "smua.nvbuffer1.readings")`

I get the timeout error. Does anyone have any idea what is going wrong here? I was able to print fine from the buffers when I ran a continuous DC sweep, so I suspect there may be something in the trigger model I've written. Happy to share any code if needed.

**Comment by amccann-Tek** [comment]
Hi @vuhoang8 Sorry for the delayed response. Do you still need help with this issue?

**Reply by vuhoang8** [reply]
Hi! Unfortunately, I am. This is what my terminal looks like when the error occurs:
```
  File ~\Environments\venv\Lib\site-packages\tm_devices\drivers\pi\pi_device.py:400 in query
    response = self._visa_resource.query(query).strip()

  File ~\Environments\venv\Lib\site-packages\pyvisa\resources\messagebased.py:647 in query
    return self.read()

  File ~\Environments\venv\Lib\site-packages\pyvisa\resources\messagebased.py:485 in read
    message = self._read_raw().decode(enco)

  File ~\Environments\venv\Lib\site-packages\pyvisa\resources\messagebased.py:441 in _read_raw
    chunk, status = self.visalib.read(self.session, size)

  File ~\Environments\venv\Lib\site-packages\pyvisa\ctwrapper\functions.py:2336 in read
    ret = library.viRead(session, buffer, count, byref(return_count))

  File ~\Environments\venv\Lib\site-packages\pyvisa\ctwrapper\highlevel.py:226 in _return_handler
    return self.handle_return_value(session, ret_value)  # type: ignore

  File ~\Environments\venv\Lib\site-packages\pyvisa\highlevel.py:251 in handle_return_value
    raise errors.VisaIOError(rv)

VisaIOError: VI_ERROR_TMO (-1073807339): Timeout expired before operation completed.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):

  File ~\Environments\venv\Lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File c:\users\peem\environments\venv\workspace\pulsed sweep\pulsed_script2.py:131
    pulsed_sweep(smu)

  File c:\users\peem\environments\venv\workspace\pulsed sweep\pulsed_script2.py:117 in pulsed_sweep
    inst.print_buffers("smua.nvbuffer2.readings", "smua.nvbuffer1.readings" )

  File ~\Environments\venv\Lib\site-packages\tm_devices\drivers\pi\tsp_device.py:172 in print_buffers
    buffer_data = self.get_buffers(*args)

  File ~\Environments\venv\Lib\site-packages\tm_devices\drivers\pi\tsp_device.py:124 in get_buffers
    buffer_str = self.query(f"printbuffer(1, {buffer_size_name}.n, {buffer_name})")

  File ~\Environments\venv\Lib\site-packages\tm_devices\drivers\pi\pi_device.py:406 in query
    raise visa.Error(msg) from error

Error: The query failed with the following message: VisaIOError('VI_ERROR_TMO (-1073807339): Timeout expired before operation completed.')
```

**Reply by amccann-Tek** [reply]
I am not an SMU expert, and as I try to find one to support in detail I think I have an idea:

And keep in mind the SMU model you are using is not fully complete.

I believe the printbuffer command you are looking for is not actually equivalent to print_buffers and might not be implemented correctly or completely at this time.

Does `inst.commands.printbuffer()` method do what you are looking for?

**Reply by vuhoang8** [reply]
Unfortunately, I get the timeout again when I use inst.commands.printbuffer() as follows:
```
    inst.commands.printbuffer(1, NUM_PULSES, "smua.nvbuffer2.readings")
    inst.commands.printbuffer(1, NUM_PULSES, "smua.nvbuffer1.readings" )
```
In this case, NUM_PULSES is set to 10. I'm wondering if I've set up the trigger model incorrectly, since inst.print_buffers() had worked when I used it in a continuous sweep script without a trigger model. At the moment, I've set up the measurement in this case like so:
```
    smua.trigger.measure.iv("smua.nvbuffer1", "smua.nvbuffer2")
    smua.trigger.measure.action = 1
    smua.trigger.measure.stimulus = period_timer.EVENT_ID
    smua.measure.rangei = 0.1
```

**Reply by amccann-Tek** [reply]
@vuhoang8 I think the best approach at this point would be to open up an [official support ticket ](https://www.tek.com/en/support/contact-technical-support)

That is where most of our support engineers reside and will get you assigned to some SMU experts. Let me know if you do not hear back from someone in a timely manner.

**Reply by vuhoang8** [reply]
Noted - thank you for your help and guidance!

**Comment by vuhoang8** [comment, marked as answer]
The issue has been resolved. In my script, I was setting attributes like so:
`smua.trigger.source.stimulus = period_timer.EVENT_ID`

However, to properly set them, you must set them as strings:
`smua.trigger.source.stimulus =  "period_timer.EVENT_ID"`

**Reply by amccann-Tek** [reply]
Thanks for following up and I am glad you got the problem sorted out.

I am the Product Manager for this package at Tek so let me know if you have any other issues, or suggestions for improving it.
